### PR TITLE
HTTP and IPFS Support

### PR DESCRIPTION
**Pull request by jbenet**
This PR adds support for http + ipfs.

## http link resolution

Support http links:

```
atom-shell app.js http://gateway.ipfs.io/ipfs/QmTKZgRNwDNZwHtJSjCp6r5FYefzpULfy37JvMt9DwvXse/video.mp4
```

Proxies traffic, with range requests.

## [ipfs](http://ipfs.io) link resolution

This PR also introduces ipfs path resolution. it tries three steps:

1. it first checks a local http gateway at localhost:8080
   (todo make this configurable), deferring to onhttplink
2. if that fails, it attempt to use the mounted filesystem.
3. if that fails, it uses the global gateway at gateway.ipfs.io.

You can try it with:

```
npm install -g go-ipfs
ipfs daemon &
sleep 10 # wait for bootstrap (this will be better)
atom-shell app.js /ipfs/QmTKZgRNwDNZwHtJSjCp6r5FYefzpULfy37JvMt9DwvXse/video.mp4
```

### CODE REVIEW

**Comment by jbenet** [comment]
@mafintosh @maxogden I resolved the divergence conflicts. this should be good to merge

File: playlist.js [hunk 1]
```diff
@@ -153,6 +154,70 @@ module.exports = function () {
...
+      stream.end()
+      return stream
+    }
+
+    // first, get the head for the content length.
+    // IMPORTANT: servers without HEAD will not work.
+    var head = file.createStream({method: 'HEAD'})
+    head.on('response', function() {
+      if (!/2\d\d/.test(head.res.statusCode))
+        cb(new Error("request failed"))
```
Comment (mafintosh) [inline playlist.js]: think you are missing a return here

File: playlist.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ var duplex = require('duplexify')
 var ytdl = require('ytdl-core')
 var events = require('events')
 var path = require('path')
+var httpStream = require('http-client-stream')
```
Comment (mafintosh) [inline playlist.js]: we already have request as a dep so i'd prefer to just use that to keep it simple if thats ok with you? :)

**Comment by mafintosh** [comment]
@jbenet other than my comments this is LGTM. btw this should also work with playback links like, playback:///ipfs/QmTKZgRNwDNZwHtJSjCp6r5FYefzpULfy37JvMt9DwvXse/video.mp4

File: playlist.js [hunk 3]
```diff
@@ -153,6 +154,70 @@ module.exports = function () {
     })
   }
 
+  var onhttplink = function (link, cb) {
+    var file = httpStream(link)
+
+    // call the http-stream's createStream with given range.
+    file.createReadStream = function(range) {
+      var rangestr = "bytes=" + range.start + '-' + range.end
```
Comment (mafintosh) [inline playlist.js]: there is actually no guarantee that `range.start` and `range.end` is set (if the request wasn't a range request) so you might want to add a if sentence here (similar to how the youtube handler handles it)

Comment (mafintosh) [inline playlist.js]: we can fix this later however

File: playlist.js (hunk 2 above)
Comment (jbenet) [inline playlist.js]: sure! am not very familiar with request; it's a large module (the readme is very long). mind handling this conversion? or pointing out what i want to do to set request headers and pull out response headers?

**Comment by jbenet** [comment]
- [ ] fix return
- [ ] use request
- [ ] handle no range

**Comment by jbenet** [comment]
@mafintosh see 26c3a16 -- might want to double check that i did the range part right.

**Comment by mafintosh** [comment]
@jbenet ranges look good

**Pull request merged**